### ISSUE TITLE
unable to start gops: unable to get current user home directory: os/user lookup failed; $HOME is empty

### ISSUE THREAD

**Issue by maxaudron** [cilium/cilium#14301]
Often times after updating cilium the hubble-ui or hubble-relay pods will fail with:
```
Error: failed to start gops agent: unable to get current user home directory: os/user lookup failed; $HOME is empty
```

Downgrading the pod to an older version sometimes works, often times not, same when upgrading and deleting the pod.

OS: ubuntu 20.10
k8s: 1.19.3
cri: cri-o with crun

**Comment by gandro** [comment]
Thanks for the report! To confirm, the error is only occurring on the `hubble-relay` pod? What version of Cilium are you using?

**Comment by maxaudron** [comment]
@gandro i've had it happen in hubble-relay and hubble-ui, but also in cilium-operator (but rarely)
currently using 1.9.1, but this has been happening on and off since 1.7

**Comment by gandro** [comment]
The `hubble-ui` backend does not contain a gops agent, as far as I can tell. @geakstr @yandzee can you confirm?

**Comment by sfxworks** [comment]
Same situation with a fresh install. The operator doesn't know what to do.

```
root@ryzen1:~# kubectl get pods -n kube-system
NAME                               READY   STATUS              RESTARTS   AGE
cilium-lrflr                       0/1     CrashLoopBackOff    4          2m34s
cilium-operator-784fcb58b7-bjg8x   0/1     Pending             0          2m34s
cilium-operator-784fcb58b7-zt9p7   0/1     CrashLoopBackOff    4          2m34s
coredns-74ff55c5b-749l6            0/1     ContainerCreating   0          10m
coredns-74ff55c5b-q2f2h            0/1     ContainerCreating   0          10m
etcd-ryzen1                        1/1     Running             0          10m
kube-apiserver-ryzen1              1/1     Running             0          10m
kube-controller-manager-ryzen1     1/1     Running             0          10m
kube-scheduler-ryzen1              1/1     Running             0          10m
root@ryzen1:~# kubectl logs -f cilium-operator-784fcb58b7-bjg8x
Error from server (NotFound): pods "cilium-operator-784fcb58b7-bjg8x" not found
root@ryzen1:~# kubectl logs -f -n kube-system cilium-operator-784fcb58b7-zt9p7
level=info msg="Skipped reading configuration file" reason="Config File \"cilium-operators\" Not Found in \"[/]\"" subsys=config
unable to start gops: unable to get current user home directory: os/user lookup failed; $HOME is emptyroot@ryzen1:~#
```

crio v1.19.3
k8s 1.20
os: ubuntu 20.04
crun version 0.16

**Comment by gandro** [comment]
Thanks for the report! So far I have been unable to reproduce this on minikube with cri-o 1.19.0. Even setting `$HOME` manually to an empty environment variable does not seem to trigger it.

But I notice both of you are running cri-o on Ubuntu 20.04. Can you share how you set up K8s? Kubeadm?

**Comment by gandro** [comment]
Looking at the source of the error, it seems like neither, `GOPS_CONFIG_DIR`, `XDG_CONFIG_HOME`, nor `HOME` are set on your particular setup, and that causes gops to error with the above error. One potential workaround would be to set `gops.Options.ConfigDir` when starting the gops agent, to avoid the failure with the missing env variables on your particular setup.

https://github.com/google/gops/blob/268f11e4fcde1604b8593e6bc35eb6e138699cb5/agent/agent.go#L85-L93
 https://github.com/google/gops/blob/268f11e4fcde1604b8593e6bc35eb6e138699cb5/internal/internal.go#L18-L42

**Comment by maxaudron** [comment]
I setup my cluster with kubeadm, cri-o from the static release bundle in the releases https://github.com/cri-o/cri-o/releases/tag/v1.19.0.

After editing the deployment and manually setting $HOME the pod started without problem.

**Comment by maxaudron** [comment]
I also have right now one operator that is failing and one that works on different nodes, but using the exact same image and cri-o versions. And on top of that the working operator is on the same node as a non-working hubble-relay.

**Comment by gandro** [comment]
I did a quick test by printing out `os.Environ()` inside the `hubble-relay` process: Both docker and cri-o on minikube set `$HOME` to `/` in the `scratch`-based `hubble-relay` image. This explains why I have not observed the issue myself so far.

So it seems that cri-o/crun on Ubuntu 20.04 for some reason does not set `$HOME` not do this. I also find it very weird that it is not happen consistently as well.

**Comment by batistein** [comment]
I have the same issue! How did you resolved it?

**Comment by gandro** [comment]
I have opened a PR which attemts to fix this on the Cilium side: #14886 - I believe however that this is probably something crun should try to address as well.

As a workaround for existing deployments, I recommend manually setting the `GOPS_CONFIG_DIR=/` environment variable on the cilium-operator and hubble-relay Kubernetes deployments.
